misk-crypto/src/main/kotlin/misk/crypto/KeyManager.kt

```diff
@@ -1,21 +1,56 @@
 package misk.crypto
 
+import com.google.crypto.tink.Aead
+import com.google.crypto.tink.Mac
+import com.google.inject.ConfigurationException
 import com.google.inject.Inject
+import com.google.inject.Injector
+import com.google.inject.Key
+import com.google.inject.name.Names
 import javax.inject.Singleton
 
-/**
- * Singleton class used as a map of key name to its corresponding [Cipher] implementation.
- */
 @Singleton
-class KeyManager @Inject constructor() {
+class AeadKeyManager @Inject internal constructor(
+  private val injector: Injector
+) {
+  private val aeads: HashMap<String, Aead> = LinkedHashMap()
 
-  private var keys: HashMap<String, Cipher> = LinkedHashMap()
+  internal operator fun set(name: String, aead: Aead) {
+    aeads[name] = aead
+  }
+
+  operator fun get(name: String): Aead? {
+    var aead = aeads[name]
+    if (aead == null) {
+      try {
+        aead = injector.getInstance(Key.get(Aead::class.java, Names.named(name)))
```

Comment:
why lazily create these instead of doing it all on startup?

Reply:
This is kind of a weird situation...
These are actually being `set()` in the `KeyManager` by their respective providers in `CryptoModule`. For example:
```
return MacFactory.getPrimitive(keysetHandle)
          .also { keyManager[key.key_name] = it }
```
And changing providers binding to `asEagerSingleton` will work and we won't need this lazy initialization.
But, it would also screw up the `InjectorTest` in every app, because when an `Aead` or `Mac` object is initialized, it needs the AWS KMS client to decrypt its key - making the `InjectorTest` test fail in every environment that's not development/testing.

By removing the `asEagerSingleton` in the module class, it means that `Aead` and `Mac` object are only initialized on first use - which is a good thing - but as a side effect also means that they won't be set in the `KeyManager` until first use *via injection*.
Which could lead to keys being missing from the `KeyManager` if they're being fetched from it on the first time.

Reply:
You could probably also multi-bind the primitives to avoid using the injector in your key managers